async-nats/src/jetstream/context.rs

```diff
@@ -930,99 +930,41 @@ impl IntoFuture for PublishAckFuture {
-    context: Context,
-    offset: usize,
-    page_request: Option<PageRequest>,
-    streams: Vec<String>,
-    done: bool,
+impl RequestSubject for String {
+    fn subject() -> String {
+        "STREAM.NAMES".to_string()
+    }
 }
 
-impl futures::Stream for StreamNames {
-    type Item = Result<String, Error>;
-
-    fn poll_next(
-        mut self: Pin<&mut Self>,
-        cx: &mut std::task::Context<'_>,
-    ) -> std::task::Poll<Option<Self::Item>> {
-        match self.page_request.as_mut() {
-            Some(page) => match page.try_poll_unpin(cx) {
-                std::task::Poll::Ready(page) => {
-                    self.page_request = None;
-                    let page = page?;
-                    if let Some(streams) = page.streams {
-                        self.offset += streams.len();
-                        self.streams = streams;
-                        if self.offset >= page.total {
-                            self.done = true;
-                        }
-                        match self.streams.pop() {
-                            Some(stream) => Poll::Ready(Some(Ok(stream))),
-                            None => Poll::Ready(None),
-                        }
-                    } else {
-                        Poll::Ready(None)
-                    }
-                }
-                std::task::Poll::Pending => std::task::Poll::Pending,
-            },
-            None => {
-                if let Some(stream) = self.streams.pop() {
-                    Poll::Ready(Some(Ok(stream)))
-                } else {
-                    if self.done {
-                        return Poll::Ready(None);
-                    }
-                    let context = self.context.clone();
-                    let offset = self.offset;
-                    self.page_request = Some(Box::pin(async move {
-                        match context
-                            .request(
-                                "STREAM.NAMES".to_string(),
-                                &json!({
-                                    "offset": offset,
-                                }),
-                            )
-                            .await?
-                        {
-                            Response::Err { error } => {
-                                Err(Box::from(std::io::Error::new(ErrorKind::Other, error)))
-                            }
-                            Response::Ok(page) => Ok(page),
-                        }
-                    }));
-                    self.poll_next(cx)
-                }
-            }
-        }
+impl RequestSubject for Info {
+    fn subject() -> String {
+        "STREAM.LIST".to_string()
     }
 }
 
-type PageInfoRequest = Pin<Box<dyn Future<Output = Result<StreamInfoPage, Error>>>>;
+trait RequestSubject {
+    fn subject() -> String;
+}
+type PageRequest<T> = Option<Pin<Box<dyn Future<Output = Result<StreamPage<T>, Error>>>>>;
 
-pub struct Streams {
+pub struct Streams<T> {
```

Comment:
Still names, since T isn't a valid stream handle?

Reply:
It doesn't have to be.
It can accept `stream::Info` or `String`, so not really `StreamNames`, isn't it?

Reply:
Hmm but the endpoint that is using this is, stream_names right? 🤔